### ISSUE TITLE
Return the same object in http request processing when session.getAttribute is called, add any "mutable" objects return from session.getAttribute to a tracker that will update Redis session key values at the end of the request processing

### ISSUE THREAD

**Issue by jchobantonov** [redisson/redisson#1868]
### Expected behavior
1) When web application gets an object from the http session, then the same object should be returned during the request processing instead of returning deserialized version of the object each time. 
2) Also each object that is "mutable" should be kept track on and at the end of the request processing should be updated to Redis server as web application could have altered the object without invoking session.setAttribute

### Actual behavior
1) When in readMode=REDIS each time session.getAttribute will return a new object instead of returning the same object returned by the first session.getAttribute during request processing.
2) When web application retrieves an object from the session and alters it then at the end of the session the Redis server won't be updated with the updated copy of the object.

### Steps to reproduce or test case
1) Try to get the object in readMode=REDIS and each time a new copy is returned during the same request processing
2) Try to update the object returned by session.getAttribute - the modified version of the attribute won't be in Redis server at the end of the request processing and the next request processing will have the old version of the object deserialized from Redis server

### Redisson version
redisson-tomcat-8-2.15.1-SNAPSHOT.jar
redisson-all-2.15.1-SNAPSHOT.jar
Tomcat 8.5.35

### Redisson configuration
<Manager className="org.redisson.tomcat.JndiRedissonSessionManager"
         readMode="REDIS"
         updateMode="DEFAULT"
         jndiName="bean/redisson" />

**Comment by jchobantonov** [comment]
@mrniko I have implemented a feature to actually cache the object from getAttibute in thread local variable (requestSessionAttributeCache) so that each time the application invokes session.getAttribute() it will get the same object - Note that in case of readMode != REDIS this is the case as we are using super.getAttribute() of Tomcat which internally uses HashMap, but in case of readMode = REDIS we are deserializing the object out of Redis server each time the application invokes session.getAttribute() - using this feature will make readMode = REDIS and readMode = MEMORY behave the same way - e.g. the object that will first be returned will be the same object if we do another session.getAttribute() this in the scope of the request processing (the only caveat here is that this is not done for String, Number, Boolean and Character as those are immutable objects).

So during request processing if the app do this using readMode = REDIS and readMode = MEMORY it will behave the same:
```
    MyObject obj = (MyObject) session.getAttribute("test");
    obj.setName("test");

    MyObject obj1 = (MyObject) session.getAttribute("test");
     
   assert(obj1.getName().equals("test")); // this should be true 

```
in both cases REDIS and MEMORY now.

_(Note that this a is a simple case where we could use the obj instead of obj1 to get the modified name but you could imagine if we have servlet Filters layered to use a common object via the session attribute then if the Filter1 is creating the obj, setting it into the session and after that modifies it, then Filter2 is invoked that will get the object out of session attribute - it will not get the same object - it will always get a clone without this feature.)_

Previously in mode REDIS this will not be true as session.getAttribute("test") will create new object out of serialized data from Redis server which deserialization could be expensive operation to do each time we call session.getAttribute() for the same object during request processing - Note that the difference between readMode=REDIS and readMode=MEMORY here is that in readMode=REDIS the same object will be returned only in the scope of the same request processing. Next request processing will get the value out of Redis server again.

The feature here also add each object that is returned from super.getAttribute() and it is not String, Number, Boolean or Character - e.g. assuming mutable object into the thread local variable as well so to keep track and update this object at the end of the request processing in order to solve following issue (and as I already mentioned Spring WebFlow is using this pattern and I guess many more web frameworks as well):

```
    private MyObject getMyObject() {
         MyObject obj = (MyObject) session.getAttribute("obj");
         if (obj == null) {
              obj = new MyObject();
              session.setAttibute("obj", obj); // without the thread local cache which will be saved at the end ot the request this will capture a snapshot of the obj that is the newly created object - no name is set here and this is what Redis server will store
         }
    }

    .....
    //somewhere during request processing
    MyObject obj = getMyObject();
    obj.setName("test");  // here the object is modified - note that with regular session implementation of Tomcat this works without issue as it is using memory HashMap - in redission one need to put readMode="MEMORY" and updateMode="AFTER_REQUEST" in order to emulate this behavior making REDIS and DEFAULT setting useless in this scenario. Using REDIS and DEFAUTL and default options should work for all web application that will otherwise work with the regular Tomcat session implementation
    .....
    
```

Key points:
In RedissonSession define:
```
    private static ThreadLocal<Map<String, Object>> requestSessionAttributeCache = new ThreadLocal<Map<String, Object>>();

    private Object getRequestSessionAttributeCacheValue(String name) {
    	Map<String, Object> localMap = requestSessionAttributeCache.get();
    	if (localMap != null) {
    		return localMap.get(name);
    	}
    	return null;
    }
    
    private void setRequestSessionAttributeCacheValue(String name, Object value) {
    	Map<String, Object> localMap = requestSessionAttributeCache.get();
    	if (localMap != null) {
    		localMap.remove(name);
    	}
    	if (value != null) {
    		if (!(value instanceof String) && !(value instanceof Number) && !(value instanceof Boolean) && !(value instanceof Character)) { // do not cache common immutable classes
    			if (localMap == null) {
    				localMap = new HashMap<String, Object>();
    				requestSessionAttributeCache.set(localMap);
    			}
    			localMap.put(name, value);
    		}
    	}
    }

    public void saveRequestSessionAttributeCache() {
    	Map<String, Object> localMap = requestSessionAttributeCache.get();
    	if (localMap != null && !localMap.isEmpty()) { // check if we gave to the application mutable objects or the application set mutable objects
    		//only store the attributes that were provided and/or stored to the application as mutable to prevent serialization of potentially big objects that were not touched by the current http request.
    		if (updateMode != UpdateMode.AFTER_REQUEST) {//check if updateMode is not AFTER_REQUEST to not store attributes into Redis twice, as after request will update all attributes at the end of the request.
    			save(localMap);
    		}
    	}
    }

    // use static method here so that if the session gets invalidated and we can't get the id out of it in RequestSessionAttributeCacheValve we should still be able to properly clean the thread
    public static void clearRequestSessionAttributeCache() {
    	requestSessionAttributeCache.remove();
    }

```

alter setAttribute, getAttribute, removeAttribute to make usage of requestSessionAttributeCache (search attached sources to see where it is used)

In RedissonSessionManager add new RequestSessionAttributeCacheValve valve in startInternal:

```
getEngine().getPipeline().addValve(new RequestSessionAttributeCacheValve());
```

And in RequestSessionAttributeCacheValve it will clear thread local variable requestSessionAttributeCache once the request processing completes and it will save any thread local session attributes that are mutable and were accesses by the application - either via setAttribute() or if getAttribute() returns "mutable" object which might be altered by the web application (this is not guarantee that the object will be modified or even mutable but this way we are going to solve the issue defined above if the web application is able to alter the object)

Here is the source of the 3 files (RedissonSession, RedissonSessionManager, RequestSessionAttributeCacheValve) in redisson-tomcat-8:

[redisson-tomcat-src.zip](https://github.com/redisson/redisson/files/2766041/redisson-tomcat-src.zip)

**Comment by jchobantonov** [comment]
@mrniko I have following concerns about thread local implementation :
1)  If web application passes the session into another thread for processing then we will have memory leak because the thread local will store attributes into the new thread map and the Valve will not be able to clear properly the thread as it won't have a reference to the new thread local - although the servlet specification forbids web application to pass http request, response and/or session object to a new different thread for processing (that is different from the http web thread) - so we should be good here, just a reminder that such thing exists

2) Because of using thread local each request processing will have its own copy of the object - e.g. if there are 2 requests each one will have it's own copy of the object from session - this is somewhat different than when we have a HashMap where the object that is returned to the 2 requests will be a shared object. I thought instead of thread local to use ConcurrentHashMap but the issue is we don't know when to clear this HashMap -  if we do it in the Valve then when the request1 completes this map should clear only the references that were made from request1 and not in request2 - and then when request2 completes only then the shared objects between request1 and request2 should be cleared - so something to consider.

**Comment by jchobantonov** [comment]
@mrniko let me know if you want me to create a pull request with the changes that I have regarding this fix (and attached as zip file)

**Comment by jchobantonov** [comment]
@mrniko 
https://github.com/redisson/redisson/pull/1871

**Comment by jchobantonov** [comment]
What do you think about this impl in PR?

**Comment by mrniko** [comment]
@jchobantonov 

Sorry, haven't reviewed it yet.

**Comment by mrniko** [comment]
> should be updated to Redis server as web application could have altered the object without invoking session.setAttribute

Seems you have some specific case. Because this statement violates the servlet specification, it requires to invoke setAttribute per object update

**Comment by mrniko** [comment]
I think this feature should be added as a new readMode. I'm going to refactor current readMode implementation in follow way: each mode will have own Session implementation object

**Comment by jchobantonov** [comment]
@mrniko 
> Seems you have some specific case. 

My case is not so specific - the issue comes from Spring WebFlow that manages navigational rules of our JSF application - so anyone who uses Spring WebFlow will have this issue - furthermore trying to get an object from the session and then immediately sticking a new object if one is not found is pretty standard practice for most of the frameworks - Sometimes you might not know when an object is changed to invoke the session.setAttribute or if you do at that place you might not have access to the HttpSession.

> Because this statement violates the servlet specification, it requires to invoke setAttribute per object update

Actually it does not. Nowhere in the servlet specification it states that once you retrieve an object from the session and you alter it you need to invoke session.setAttribute method for this object.

From Servlet Specification:
```
Distributed Environments
Within an application marked as distributable, all requests that are part of a session
must be handled by one JVM at a time. The container must be able to handle all
objects placed into instances of the HttpSession class using the setAttribute or
putValue methods appropriately. The following restrictions are imposed to meet
these conditions:
1) The container must accept objects that implement the Serializable interface.
2) The container may choose to support storage of other designated objects in the
HttpSession, such as references to Enterprise JavaBeans components and
transactions.
3) Migration of sessions will be handled by container-specific facilities.
```

As you could see from the above even Enterprise JavaBeans could be stick into the session - now you don't know when a EJB could change it's state if it is session EJB bean to also try and alter the HttpSession using setAttribute. For example if the EJB session bean participate in transaction it could listen for transaction completion and alter its's state - from there you don't have access to HttpSession and you should not as the EJB might not even support dealing with HttpSession.
In the session it is just a handle to the EJB not the serialized version of the actual object - so if you could have a handle to any object you stick into the session - then yes - you could actually update the handle only using setAttribute - then each time you deserialize the handle you will then retrieve from the handle the actual object - the same case with java.io.File - it is a handle to content - if you stick the file into the session only the handle will be stored then when you retrieve the object the handle will be deserialized but the actual file content could have been changed already in the file system - the same concept here - the only thing is that you do not have a handle but the actual object and this actual object needs to be serialized

Note that our application, using Spring WebFlow, works just fine with standard session implementation - as the standard implementation is using regular java Map - no issues there - (even when you do session.getAttribute() it will return the same object for the entire request processing - this is not the case with Redis read mode where you return a new copy out of deserialization each time - at least changing identity of the object - the equals method could work but the ==  operation will not - e.g session.getAttribute("myObject) != session.getAttribute("myObject) will be true)- and as soon as we switch to Redis read mode session implementation it will complain that there is missing Spring WebFlow conversation because each time getAttribute will return deserialized version of the object that is not populated at all when session.setAttribute is called.

>I think this feature should be added as a new readMode. 

Thanks. Although I think your default implementation should be as such that it will track objects and update them into the session
What you could do is to implement your own ObjectHandle if you want - like this:

```
package org.redisson.tomcat;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectOutputStream;
import java.io.Serializable;

import org.apache.catalina.util.CustomObjectInputStream;

public class ObjectHandle implements Serializable {
	 private static final long serialVersionUID = 3131004535786310056L;
	 private byte[] actualObjectAsBytes;
	 private transient Object actualObject;
	 
	 public Object getActualObject(ClassLoader classLoader) throws IOException, ClassNotFoundException {
		if (actualObject == null) {
			if (actualObjectAsBytes != null) {
				actualObject = toObject(classLoader, actualObjectAsBytes);
			}
		}
		return actualObject;
	 }

	 public void setActualObject(Object actualObject) throws IOException {
		this.actualObject = actualObject;
		this.actualObjectAsBytes = toByteArray(actualObject);
	 }

	 protected byte[] toByteArray(Object value) throws IOException {
		if (value == null) {
			return null;
		}
		ByteArrayOutputStream bos = new ByteArrayOutputStream();
		ObjectOutputStream out = new ObjectOutputStream(bos);
		out.writeObject(value);
		out.flush();
		return bos.toByteArray();
	 }
	
	 protected Object toObject(ClassLoader classLoader, byte[] value) throws IOException, ClassNotFoundException {
    	if (value == null) {
    		return null;
    	}
    	CustomObjectInputStream in = new CustomObjectInputStream(new ByteArrayInputStream(value), classLoader);
    	return in.readObject();		
	 }
}

```

So basically you are going to store a byte array into Redis - this ObjectHandler will help you with cases where you can't deserialize an object - so to throw exception only if the application invokes session.getAttribute("myAttribute") and myAttribute can't be deserialized - but if the request do not call session.getAttribute("myAttribute") - there is no need to actually throw exception when the redisson map can't be created because of serialization issues - and you will know the correct classloader will be used - Now you need to have internal HashMap for the duration of the request 

1) request comes
2) application invokes session.getAttribute
3) you check internal HashMap if you do have ObjectHandle - if not then use redis map to get one and then use this object each time during request processing 
4) when session.setAttribute is invoked you will change the object handle to point to the other object - but at the end of the request you should have Tomcat Valve that will clear this internal Map and store all ObjectHandle into Redis - now if the object has been changed - then the ObjectHandle will serialize the actual object as when you call ObjectHandler.getActualObject the actualObject will not be null and it will return what you need

**Comment by jchobantonov** [comment]
@mrniko 
also if the user of the redisson-tomcat do not know what he/she is doing it could first set one codec and then set it to another one - obviously any sessions currently living in redis will not be able to be deserialized - also no one in servlet specification says that the user could choose it's own codec implementation :) - only that it should implement Serializable which is a great hint that serialization will be used and not some other codec implementations
Although in servlet specification it is saying that objects could not expect its readObject, writeObject to be invoked during serialization/deserialization there could be some frameworks that could use this feature to alter its own serialization.

From servlet specification:

```
If distributed containers persist or migrate sessions to provide quality of service
features, they are not restricted to using the native JVM Serialization mechanism for
serializing HttpSessions and their attributes. Developers are not guaranteed that
containers will call readObject and writeObject methods on session attributes if
they implement them, but are guaranteed that the Serializable closure of their
attributes will be preserved. 
```
The important thing here is 
>but are guaranteed that the Serializable closure of their
attributes will be preserved

Which means that other codecs should not be used as there is no guarantee that they will preserve the state of the object correctly.

also note this from the servlet specification:
```
The distributed servlet container must throw an IllegalArgumentException for
objects where the container cannot support the mechanism necessary for migration
of the session storing them
```
So if you try to serialize an object then IllegalArgumentException should be thrown - e.g. wrapping any ClassNotFoundException etc. Oddly enough you could get this exception even when you try to write - here is an exception that happen for me during local debugging (still not sure how this happen and if it will happen again - yet the exception is experienced by other as well):

```
java.lang.ClassNotFoundException: org.richfaces.function.RichFunction
        at java.lang.Class.forNameImpl(Native Method)
        at java.lang.Class.forName(Class.java:365)
        at org.apache.el.util.ReflectionUtil.forName(ReflectionUtil.java:65)
        at org.apache.el.lang.FunctionMapperImpl$Function.getMethod(FunctionMapperImpl.java:158)
        at org.apache.el.lang.FunctionMapperImpl$Function.writeExternal(FunctionMapperImpl.java:125)
        at java.io.ObjectOutputStream.writeExternalData(ObjectOutputStream.java:1526)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1497)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at java.util.concurrent.ConcurrentHashMap.writeObject(ConcurrentHashMap.java:1424)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1200)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1563)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at org.apache.el.lang.FunctionMapperImpl.writeExternal(FunctionMapperImpl.java:75)
        at java.io.ObjectOutputStream.writeExternalData(ObjectOutputStream.java:1526)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1497)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at org.apache.el.ValueExpressionImpl.writeExternal(ValueExpressionImpl.java:272)
        at java.io.ObjectOutputStream.writeExternalData(ObjectOutputStream.java:1526)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1497)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at com.sun.facelets.el.TagValueExpression.writeExternal(TagValueExpression.java:128)
        at java.io.ObjectOutputStream.writeExternalData(ObjectOutputStream.java:1526)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1497)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at java.util.HashMap.internalWriteEntries(HashMap.java:1801)
        at java.util.HashMap.writeObject(HashMap.java:1374)
        at sun.reflect.GeneratedMethodAccessor407.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1200)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1563)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at java.util.ArrayList.writeObject(ArrayList.java:777)
        at sun.reflect.GeneratedMethodAccessor479.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1200)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1563)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at java.util.ArrayList.writeObject(ArrayList.java:777)
        at sun.reflect.GeneratedMethodAccessor479.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1200)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1563)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at java.util.ArrayList.writeObject(ArrayList.java:777)
        at sun.reflect.GeneratedMethodAccessor479.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1200)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1563)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1615)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1576)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at java.util.LinkedHashMap.internalWriteEntries(LinkedHashMap.java:344)
        at java.util.HashMap.writeObject(HashMap.java:1374)
        at sun.reflect.GeneratedMethodAccessor407.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1200)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1563)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at java.util.LinkedHashMap.internalWriteEntries(LinkedHashMap.java:344)
        at java.util.HashMap.writeObject(HashMap.java:1374)
        at sun.reflect.GeneratedMethodAccessor407.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1200)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1563)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1615)
        at java.io.ObjectOutputStream.defaultWriteObject(ObjectOutputStream.java:458)
        at org.ajax4jsf.application.AjaxStateHolder.writeObject(AjaxStateHolder.java:197)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1200)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1563)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at org.redisson.codec.SerializationCodec$2.encode(SerializationCodec.java:67)
        at org.redisson.RedissonObject.encodeMapValue(RedissonObject.java:278)
        at org.redisson.RedissonMap.putAllOperationAsync(RedissonMap.java:424)
        at org.redisson.RedissonMap.putAllAsync(RedissonMap.java:365)
        at org.redisson.RedissonMap.putAll(RedissonMap.java:296)
        at org.redisson.tomcat.RedissonSession.save(RedissonSession.java:347)
        at org.redisson.tomcat.RedissonSession.saveRequestSessionAttributeCache(RedissonSession.java:111)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:52)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:342)
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:800)
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:806)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1498)
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1160)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:811)
```

**Comment by jchobantonov** [comment]
@mrniko any comment on what I put ?

**Comment by jchobantonov** [comment]
@mrniko here is a strange ClassNotFoundException that I got during session attribute serialization - not when we try to deserialize the bean:

```

java.lang.ClassNotFoundException: org.richfaces.function.RichFunction
        at java.lang.Class.forNameImpl(Native Method)
        at java.lang.Class.forName(Class.java:402)
        at org.apache.el.util.ReflectionUtil.forName(ReflectionUtil.java:65)
        at org.apache.el.lang.FunctionMapperImpl$Function.getMethod(FunctionMapperImpl.java:158)
        at org.apache.el.lang.FunctionMapperImpl$Function.writeExternal(FunctionMapperImpl.java:125)
        at java.io.ObjectOutputStream.writeExternalData(ObjectOutputStream.java:1526)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1497)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at java.util.concurrent.ConcurrentHashMap.writeObject(ConcurrentHashMap.java:1424)
        at sun.reflect.GeneratedMethodAccessor777.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1212)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1563)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at org.apache.el.lang.FunctionMapperImpl.writeExternal(FunctionMapperImpl.java:75)
        at java.io.ObjectOutputStream.writeExternalData(ObjectOutputStream.java:1526)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1497)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at org.apache.el.ValueExpressionImpl.writeExternal(ValueExpressionImpl.java:272)
        at java.io.ObjectOutputStream.writeExternalData(ObjectOutputStream.java:1526)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1497)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at com.sun.facelets.el.TagValueExpression.writeExternal(TagValueExpression.java:128)
        at java.io.ObjectOutputStream.writeExternalData(ObjectOutputStream.java:1526)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1497)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at java.util.HashMap.internalWriteEntries(HashMap.java:1801)
        at java.util.HashMap.writeObject(HashMap.java:1374)
        at sun.reflect.GeneratedMethodAccessor584.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1212)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1563)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at java.util.ArrayList.writeObject(ArrayList.java:777)
        at sun.reflect.GeneratedMethodAccessor585.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1212)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1563)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at java.util.ArrayList.writeObject(ArrayList.java:777)
        at sun.reflect.GeneratedMethodAccessor585.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1212)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1563)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at java.util.ArrayList.writeObject(ArrayList.java:777)
        at sun.reflect.GeneratedMethodAccessor585.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1212)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1563)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1431)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1205)
        at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1615)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1576)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at java.util.LinkedHashMap.internalWriteEntries(LinkedHashMap.java:344)
        at java.util.HashMap.writeObject(HashMap.java:1374)
        at sun.reflect.GeneratedMethodAccessor584.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1212)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1563)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at java.util.LinkedHashMap.internalWriteEntries(LinkedHashMap.java:344)
        at java.util.HashMap.writeObject(HashMap.java:1374)
        at sun.reflect.GeneratedMethodAccessor584.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1212)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1563)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1615)
        at java.io.ObjectOutputStream.defaultWriteObject(ObjectOutputStream.java:458)
        at org.ajax4jsf.application.AjaxStateHolder.writeObject(AjaxStateHolder.java:197)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1212)
        at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1563)
        at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1499)
        at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1209)
        at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:365)
        at org.redisson.codec.SerializationCodec$2.encode(SerializationCodec.java:67)
        at org.redisson.RedissonObject.encodeMapValue(RedissonObject.java:278)
        at org.redisson.RedissonMap.putAllOperationAsync(RedissonMap.java:424)
        at org.redisson.RedissonMap.putAllAsync(RedissonMap.java:365)
        at org.redisson.RedissonMap.putAll(RedissonMap.java:296)
        at org.redisson.tomcat.RedissonSession.save(RedissonSession.java:347)
        at org.redisson.tomcat.RedissonSession.saveRequestSessionAttributeCache(RedissonSession.java:111)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:52)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.redisson.tomcat.RequestSessionAttributeCacheValve.invoke(RequestSessionAttributeCacheValve.java:45)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:342)
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:800)
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:806)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1498)
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1160)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:812)

```

it seems that the valve is not executed inside application thread classloader and forClass fails
Here is the modified versions of RedissonSessionManager, and RequestSessionAttributeCacheValve and UpdateValve to solve this Thread context classloader issue using Thread.currentThread(),setContextClassLoader() method. Also the Valves are removed from the pipeline not just added and we have a request attribute to check if the request has been already filterer or not

[tomcat.zip](https://github.com/redisson/redisson/files/2857547/tomcat.zip)

**Comment by jchobantonov** [comment]
[tomcat.zip](https://github.com/redisson/redisson/files/2857689/tomcat.zip)
here is updated tomcat.zip that will put the context classloader only when we try to do save for the session not on entire body of the Valve

**Comment by mrniko** [comment]
I see too many invocations of `RequestSessionAttributeCacheValve.invoke` in log, seems it's invoked recursively.

**Comment by jchobantonov** [comment]
@mrniko the RequestSessionAttributeCacheValve is a new class in order to support storage of the attributes that were returned to the application (in case the application modifies them) at the end of the request - so no it can't be reproduced with the master branch as you do not have this capability or valve.

The too many invocations of RequestSessionAttributeCacheValve.invoke is not because it's invoked recursively - all valves and filters could be executed multiple times in one request - that's because the request could have includes, redirects etc where valves and filters also participates - in order to apply the filter only once I have used request attribute (ALREADY_FILTERED_ATTRIBUTE) - so check if the attribute is present - if so just invoke getNext().invoke(request, response) - if the attribute is not present then execute the filter logic - you could find the same pattern used by Spring's OncePerRequestFilter class. Note that in the stack trace RequestSessionAttributeCacheValve.invoke could still be seen multiple times - this attribute do not prevent this - it only prevents applying the logic of the filter multiple times per request.

I already saw what is the issue - the issue is that the save in  RequestSessionAttributeCacheValve is executed outside (web) application classloader - that't why when the ObjectOutputStream tries to write the object it will fail as the reflection API that apache EL is using is going to be executed outside the web application classloader - note that all Tomcat valves are executed outside application classloader - I have tried to fix that by passing the classloader that we got from RedissonSessionManager using this:
```
        final ClassLoader applicationClassLoader;
        if (Thread.currentThread().getContextClassLoader() != null) {
            applicationClassLoader = Thread.currentThread().getContextClassLoader();
        } else {
            applicationClassLoader = getClass().getClassLoader();
        }
```
but it turns out that when RedissonSessionManager.startInternal is invoked the thread do not have the application classloader in the thread context classloader - that's why even when I tried to pass this classloader into the valve it blows as well with ClassNotFoundException

In order to solve this I had to do this in the valve (note I have applied the same in UpldateValve as well as this is an issue there as well as the save is executed outside web application context/classloader):

```
    final ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
    try {
        ClassLoader applicationClassLoader = request.getContext().getLoader().getClassLoader();
        Thread.currentThread().setContextClassLoader(applicationClassLoader);
        //... here do session.save etc that will need the web application classloader in order to serialize the data - as apache EL implementation will use Class.forName with a classloader out from Thread.currentThread().getContextClassLoader() we are fine here
    } finally {
       Thread.currentThread().setContextClassLoader(classLoader);
    }
```

I have fixed the new RequestSessionAttributeCacheValve, UpdateValve and RedissonSessionManager to use getContext().getLoader().getClassLoader() to be passed into updatesTopic.addListener

I'm attaching all modified classes that are up to date with master branch so you could copy those files into your local master branch and do compare to see the diff.

Both RequestSessionAttributeCacheValve and UpdateValve  are now stored as variables inside RedissonSessionManager  so that when stopInternal is invoked it could properly remove the valves that are added to the engine pipeline.

[tomcat.zip](https://github.com/redisson/redisson/files/2861380/tomcat.zip)

Also one question - I saw that latest in master requires now java 1.8 - are you dropping support for java 7 with the latest redisson release?

**Comment by jchobantonov** [comment]
@mrniko btw with the above changed files I do not see this exception anymore - note that I have not tested out the changes to RedissonSessionManager classloader as I'm not using this configuration to update the session attributes when there is update in redisson - but I have already verified that the applicationClassLoader variable that have been initialized in RedissonSessionManager.startInternal do not work as it is not the web application classloader

**Comment by jchobantonov** [comment]
@mrniko  here are the files that I had to change to implement this and they are aligned with the master branch as of today so you could take a look at the diff:
[redisson-session.zip](https://github.com/redisson/redisson/files/3112851/redisson-session.zip)

**Comment by jchobantonov** [comment]
there are 3 files in above redisson-session.zip file: RedissonSession.java (modified), RedissonSessionManager.java (modified), RequestSessionAttributeCacheValve.java (new)

**Comment by mrniko** [comment]
@jchobantonov 

Will have a look. Thank you!

**Comment by mrniko** [comment]
Sorry, could you clarify why can't we use ReadMode.MEMORY? Seems it does the same - stores session state in internal map

**Comment by jchobantonov** [comment]
First thing first - The reason I select ReadMode.REDIS over ReadMode.MEMORY is because it will get a fresh copy of the session from Redis each time at the beginning of the request - the issue with ReadMode.MEMORY is that if there is network issue (been unavailable for some time on one of the application nodes) and during this time an event is send to synchronize the session - like new attribute is been added - we could end up with a stale session on this node - the ReadMode.MEMORY will only work right if we could guarantee there is no issue with topic event notification - also there is another issue with ReadMode.MEMORY - as soon as something happens on one node in the session the event will be send and all nodes will see this immediate change in the middle of the request process - so if a request is in process and we ask for an attribute it could have one value at the beginning of the request processing and in the middle of the request to have another one because another node did a change using different request on different node and the event broadcast updated the internal memory state.

Now as to the reason why ReadMode.MEMORY also will not work without fixing this issue. Here is the scenario:

1) We have 2 application nodes (servers)
2) On server 1 we process request and do this:
```
MyBean myBean = new MyBean();
session.setAttribute("myBean", myBean);
myBean.setName("Test");
```
3) Now on server 1 we have in the session myBean as attribute and the bean is with name "Test" as we are using RedisMode.MEMORY
4) The server 2 will now have also myBean but it will not have the name "Test" and it will never have this name as there is no one to synchronize the myBean.setName("Test") change on the myBean - the synchronization happens only during setAttribute - after that any change to the bean will only stay on server 1

Now Tomcat internal session management is using a Map so such things like to stick something to the session and then modify it without actually calling session.setAttribute are quite common practice for some of the frameworks around - like as I mentioned Spring WebFlow JSF, and they work as Tomcat is using a reference when we stick a bean into the session with setAttribute method

So my suggestion is that any object that we return from the session or setting an object into the session that is mutable we keep track and then at the end of the request we update all of those objects as we don't know if we have changed the internal state of the object after the object is been obtained from the session or we set it using setAttribute

so any object that is been retrieved using :
getAttribute()
or been set to the session using 
setAttribute()
we add it to be tracked

Now this is different than update mode AFTER_REQUEST as we only track objects that were accessed 
from the current request processing - we could have like 100 objects in the session but if the current request processing do not touch any of them - e.g. do not call getAttribute or setAttribute then we will not update anything while AFTER_REQUEST will update all objects in the session again and it will do that for all objects - not only for the one that were touched.

By saying mutable here I refer to the checks in the code that you could see like if the object that we get from getAttribute or setting it into the session with setAttribute() is a primitive type or a wrapper class of it we know those are immutable and there is no need to track them - as there are a lot of Strings etc been put into the session - now because we don't know what the other cases are we are just going to update the Redis with them - there could be beans that are not mutable - like do not have set methods in them but because in java we don't have some checks to tell us if the bean is immutable or if it is mutable but have not changed it's internal state we update those at the end of the request.

**Comment by jchobantonov** [comment]
@mrniko ^

**Comment by mrniko** [comment]
> as soon as something happens on one node in the session the event will be send and all nodes will see this immediate change in the middle of the request process - so if a request is in process and we ask for an attribute it could have one value at the beginning of the request processing and in the middle of the request to have another one because another node did a change using different request on different node and the event broadcast updated the internal memory state

This could be solved with `updateMode = AFTER_REQUEST`

**Comment by jchobantonov** [comment]
@mrniko about : 

> This could be solved with updateMode = AFTER_REQUEST

There are 2 issues with that
1) introducing performance issue - so basically if you do have 100 rest end points (servlets) and only one of them under certain conditions uses the session by setting updateMode = AFTER_REQUEST you will force session persistence for all 100 rest end points no matter that only one of them in it's implementation uses the session. More over if the http session is pretty big like 100 object graphs and we did not touch any of them with the current request processing, or only a few of them, we will serialize all of the 100 object graphs at the end of each request no matter what - so we have serialization impact per each request as well as network transfer/connect to Redis for the entire session 

2) introducing concurrency issue - let say user sends 2 different concurrent requests to 2 different servlets that touch different session attributes. Because at the end we will force each request to store the entire session into redis only one set of changes will end up in redis - not like the merge of both - and sending 2 different request to the web server is pretty common - like if you put 2 URL in a web page that are let say images produced by servlet then you will have concurrent requests on your hand. And as I mentioned previously if ReadMode.MEMORY is been used and server 2 disconnects for a little bit from the network and it misses attribute update then the in memory session will be stale, so after that any request to the server 2 will override Redis server to have the stale version of the objects no matter that the request that is been process could not touch the stale value in the session

**Comment by jchobantonov** [comment]
So if you are willing to change the behavior  of AFTER_REQUEST to only update touched objects from the session I’m fine with that and I will configure my app to use after request update

**Comment by jchobantonov** [comment]
just FYI - if you want to implement this using AFTER_REQUEST then you need also to fix the UpdateValve to have request attribute marker so the UpdateValve is only executed once at the end of the request - just like the attached RequestSessionAttributeCacheValve that I have put here - see the ALREADY_FILTERED_ATTRIBUTE request attribute. Having such attribute will prevent the valve to save the session objects multiple times if we have request that in turn do include etc as the Valves will be executed per each include/forward as well - so using request attribute marker will guarantee that the valve is executed only once per request processing (instead of multiple times if the request processing have include/forward to another servlet/jsp to finish the processing of the request)

**Comment by jchobantonov** [comment]
@mrniko ^

**Comment by jchobantonov** [comment]
and one more thing - the attached classes also solves the problem of not deserializing the object each time from redis in case the readMode = REDIS - it will be done once per request so following will be true:
```
Object myObject = session.getAttribute("myObject");
....
if (myObject == session.getAttribute("myObject")) {
    // the execution will go here as the session.getAttribute("myObject") is cached 
    // in request thread local cache and session.getAttribute("myObject") will not trigger 
    // retrieval and/or deserialization of the object again and returning a new object
}
```
so this and the fact that all touched objects will be saved at the end of the request is making the feature close to Tomcat's usage of Map - e.g. if you put some object into a map you will expect that the same object is returned (by identity) if you call getAttribute() and if you change the object like this:
```
MyObject myObject = session.getAttribute("myObject");
myObject.setName("test");
```
or like this:
```
MyObject myObject = session.getAttribute("myObject");
if (myObject == null) {
   myObject = new MyObject();
   session.setAttribute("myObject", myObject);
}
myObject.setName("test");
```
it will be persisted across different requests - so if the application works with Tomcat's default session manager it should work with Redisson session manager as well - and as I already pointed out a web application that works perfectly fine with Tomcat's default session manager are not working with Redisson as this issue prevents that. I think by default Redisson should behave close to the semantic of hashmap as this is what tomcat by default it is using

**Comment by jchobantonov** [comment]
I wanted also to point out that instead of request.setAttribute() / getAttribute() for UpdateValve to check if it was already executed we could use setNote / getNote(java.lang.String name) for the request - so that the web application will not see the request attribute if it tries to list all request attributes - also the notes could be used for keeping request local cache instead of using Thread local variable - the request notes as cache it is better in cases we are dealing with async servlets where the async servlet using threadlocal cache will set the cache into servlet thread and the filter finally block could be executed in the original request thread which will not be able to clean properly the thread local variable so I guess using notes here will be better approach

**Comment by jchobantonov** [comment]
@mrniko here are the changes that I'm using right now against latest master:
[tomcat.zip](https://github.com/redisson/redisson/files/3291449/tomcat.zip) - those are for tomcat-8

The first attempt was to add a thread local hashmap in order to return the same copy of an attribute object over and over again during request processing - this helped with spring webflow and JSF as there is a code like this:
```
Object obj = session.getAttribute("abc");
if (obj == null) {
   obj = .... //create object;
   session.setAttribute("abc", obj);
}
obj.setName("test");
```
... further down to the request processing (in another filter/servlet) 
```
Object obj = session.getAttribute("abc")
```

FYI the exception thrown from spring webflow was that there is no conversation "1" found as the conversation object that was put into the session was only created and then put into the session immediately and only after that it was modified to have the conversation object "1" but at the end of the processing when it tried to find this conversation from the session it wasn't able to as the returned object from the session was not initialized at all.

the expectation is that obj will be the same object as previously stored in the session but with current REDIS read mode in redisson master branch it will not be - it will be a copy of deserialized object that even do not have the name set - as obj.setName is invoked after we store the attribute into the session - e.g. the object is modified when it was returned from the session via getAttribute()/setAttribute()

For this I used threadlocal hashmap that will store an attribute cache and it will track each object that was return via session.getAttribute() so that after the request is processed it will store the possibly modified object back into redis, as well it will give you the same object over and over again instead of returning each time a copy of the object

Down the road thought our app experienced concurrency issue - as you might know JSF is stateful and Spring WebFlow helps with the transitions as a state machine - what we have noticed was that the state on the server was somewhat corrupted - e.g. if you have 2 concurrent request at the end the state on the server will be like only one happen - e.g. the objects inside JSF state will have some properties not set from one of the requests

Then I saw a code in spring webflow like this:
 org.springframework.webflow.conversation.impl.SessionBindingConversationManager:
```
	private ConversationContainer getConversationContainer() {
		SharedAttributeMap sessionMap = ExternalContextHolder.getExternalContext().getSessionMap();
		synchronized (sessionMap.getMutex()) {
			ConversationContainer container = (ConversationContainer) sessionMap.get(sessionKey);
			if (container == null) {
				container = createConversationContainer();
				sessionMap.put(sessionKey, container);
			}
			return container;
		}
	}
```
Now the Mutex object is retrieved by this class:
org.springframework.webflow.context.servlet.HttpSessionMap:

```
	public Object getMutex() {
		// force session creation
		HttpSession session = request.getSession(true);
		Object mutex = session.getAttribute(WebUtils.SESSION_MUTEX_ATTRIBUTE);
		return mutex != null ? mutex : session;
	}
```

During my testing the mute object above was null so we are going to synchronize by the http session itself - but you could see easily here that some frameworks could use java synchronization on session attribute object. 

More over if there are 2 concurrent threads accessing the session at the same time if I use per thread session attribute cache then the 2 concurrent request processing threads will each have it's own copy of the objects been deserialized from the session and at the end only one of them will succeed to update the redis session - e.g. the other one will override the changes made by the first concurrent request that was completed.

In order to fix this I did the following:
Create a new hashmap that will share the objects between all threads - just like tomcat default session hashmap is doing, but put which thread is using this object - e.g. when a thread asks about session attribute put that this thread is using this attribute so that once the RequestSessionAttributeCacheValve completes to remove the cache object and store it into Redis only if this object is used only by this thread - if more threads are using this then leave the object in the cache and remove that the object is used by this thread - at the end the object will be removed from the cache when all threads using this session attribute has completed it's request processing and at that time we will store the object into the redis server (it is been stored at the spot if setAttribute is invoked as well...)

So far I do not see concurrency issue coming from our testing

And again - I'm using REDIS read mode instead of MEMORY because I want to guarantee that if there are some disconnect between redisson and redis server we do not loose update events and this makes MEMORY cache stale - and by having like 20-30 web applications in one tomcat instance we are not going to create too many connections to redis as I think each tomcat app will create it's own topic to publish events etc and it's own connection - having too many web apps will create too many network connections which could be problematic and I have seen in kubernetes aws cluster that we are using sometimes redisson have some exceptions coming from dns server that it can't resolve the redis host so I guess topic connections could have been lost at some point making stale sessions in other tomcat nodes. - this is the reason to use REDIS read mode instead of MEMORY

As for UpdateMode.AFTER_REQUEST mode as I have explained this do not work fine if you have 2 requests that could hit 2 different nodes - like request1 hit node1 and request2 hit node2 - then if request1 update attribute1 and request2 update attribute2 at the end only one of them will succeed as the after request will update all session attributes instead of just the one that the current request have used (an effective merge of the session attributes that were modified) - e.g. the attributes that were obtained via getAttribute() and modified after that, used in setAttribute or removeAttribute

As you might have notice RequestSessionAttributeCacheValve will execute 
```
((RedissonSession) sess).saveRequestSessionAttributeCache(Thread.currentThread()); 
```
each time as we want to pass the thread into the session that was used for this request part processing as there could be more than one thread involved in fully processing of the request and that's the reason we can't use ALREADY_FILTERED_NOTE there as we want to pass which is the thread used to obtain session attribute and clear it's usage at the end

There is one issue about using this approach and that's with async requests as when a request comes it could be started up for processing in another thread and the RequestSessionAttributeCacheValve will not be able to properly clean which threads do have access to the attribute so that part is not done - it have a WeakReference to the thread so if the async thread is discarded after that it will "fix" the leak but that's not guarantee and a proper fix should be added or some other ways to track the usage of session attributes should be employed. 

But the idea is this - keep track of attributes that were used during request processing and update them at the end of the request processing if they are not used by other request processing that are currently happening.

**Comment by jchobantonov** [comment]
After some more testing this seems to be working:
[tomcat.zip](https://github.com/redisson/redisson/files/3303694/tomcat.zip)

the previous one did not work correctly as RequestSessionAttributeCacheValve has been installed multiple times into tomcat pipeline (once per each web app) which is causing concurrency issues - e.g. premature removal from the attribute cache which is causing invalid JSF state - as RequestSessionAttributeCacheValve is not using the pattern ALREADY_FILTERED_NOTE as we need to know which is the thread that is processing the request so that we could remove session attributes that were used by this thread - so each filter should be considered - but the expectation is that RequestSessionAttributeCacheValve  is installed only once - however because we are not tracking if the valve is installed to not install it again it brings the issue.

**Comment by jchobantonov** [comment]
@mrniko here is PR for this issue for tomcat 8 : https://github.com/redisson/redisson/pull/2194

**Comment by jchobantonov** [comment]
@mrniko 

Here are the 3 classes in redisson-tomcat-8 project that I modified against today's latest redisson snapshot:
RedissonSession
RedissonSessionManager
UpdateValve

[tomcat.zip](https://github.com/redisson/redisson/files/4019758/tomcat.zip)

The implementation is using local in memory cache - e..g when a request comes it will get the object from redisson (deserialization) but during the request processing it will give you the same object over and over again - it will not try to deserialize the object again from redis - at the end UpdateValve will check if the local in memory cache is used by other Requests and if not it will serialize them into redis and discard the in memory cache - if during the processing of the request another request comes and uses the in memory cache then the first request will not discard anything but at the end of the second request we will serialize those in memory caches into redis and discard the in memory cache so that the next time we will obtain potentially changed objects from redis

Using in memory cache will prevent session.getAttribute to get a new object each time - this preserving object identity that could be used in IdentityMap during request processing - in the case of JSF if they are 2 concurrent requests we need to have the same objects from the session.getAttribute - e.g. to share the objects and the JSF to be able to synchronize those request using objects in the session - if each request get's it's own object this will not be possible

**Comment by mrniko** [comment]
@jchobantonov 

Thanks for proposed changes. I think this feature should be implemented as a new mode, like REDIS_INSTANCE_CACHE or something.

**Comment by mrniko** [comment]
@jchobantonov 

I have fixed https://github.com/redisson/redisson/issues/2230. Will that resolve this issue as well?

**Comment by jchobantonov** [comment]
@mrniko sorry for the late response, I just saw the fix #2230 - it does fix some of the issues but the issue that still is not resolved I think is when the cache updatedAttributes should be cleared - right now the updatedAttributes  is cleared when save is invoked which is happening at the end of the request via a valve. The problem is that the session is shared among all of the request - so if we have this situation:
begin request1 -> update/create attribute1
begin request2 -> update/create attribute2
end request1 (here the save will happen and clear the updateAttributes)
any request2 accessign attribute2/1 will result in fetch from redis which will create new object which would be different than the original

what my PR is doing is tracking who is using the session and when the valve end invokes save it will only clear the cache if there are no other requests that are in place
so 
begin request1
begin request2
end request1 (this will not clear the cache)
end request 2 (there are not other request so the clear will happen and save will occur)

**Comment by mrniko** [comment]
Fixed in https://github.com/redisson/redisson/issues/3032
